crates/runtime/src/traphandlers.rs

```diff
@@ -591,45 +592,86 @@ mod tls {
 
     /// Opaque state used to help control TLS state across stack switches for
     /// async support.
-    pub struct TlsRestore {
+    pub struct TlsState {
+        // The head of a linked list, similar to the TLS state. Note though that
+        // this list is stored in reverse order to assist with `push` and `pop`
+        // below.
+        //
+        // After a `push` call this stores the previous head for the current
+        // thread so we know when to stop popping during a `pop`.
         state: raw::Ptr,
     }
 
-    impl TlsRestore {
-        /// Takes the TLS state that is currently configured and returns a
-        /// token that is used to replace it later.
-        ///
-        /// This is not a safe operation since it's intended to only be used
-        /// with stack switching found with fibers and async wasmtime.
-        pub unsafe fn take() -> TlsRestore {
-            // Our tls pointer must be set at this time, and it must not be
-            // null. We need to restore the previous pointer since we're
-            // removing ourselves from the call-stack, and in the process we
-            // null out our own previous field for safety in case it's
-            // accidentally used later.
-            let state = raw::get();
-            if let Some(state) = state.as_ref() {
-                state.pop();
-            } else {
-                // Null case: we aren't in a wasm context, so theres no tls to
-                // save for restoration.
+    impl TlsState {
+        /// Creates new state that initially starts as null.
+        pub fn new() -> TlsState {
+            TlsState {
+                state: std::ptr::null_mut(),
             }
-
-            TlsRestore { state }
         }
 
-        /// Restores a previous tls state back into this thread's TLS.
+        /// Pushes the list stored in this state onto the current thread's
+        /// state.
+        ///
+        /// This will iterate over the linked list of states stored within
+        /// `self` and push them sequentially onto the current thread's
+        /// activation list.
         ///
-        /// This is unsafe because it's intended to only be used within the
-        /// context of stack switching within wasmtime.
-        pub unsafe fn replace(self) {
-            if let Some(state) = self.state.as_ref() {
+        /// # Unsafety
+        ///
+        /// Must be carefully coordinated with `pop` and fiber switches to
+        /// ensure that this doesn't push stale data and the data is popped
+        /// appropriately.
+        pub unsafe fn push(&mut self) {
```

Comment:
So currently nothing is really stopping us from calling `push` twice, with no `pop`s in between, which would lead to broken things. Do you think it is worth having two types here? 

1. a `PushedTlsState` for when it is pushed, and this would have a `fn pop(self) -> PoppedTlsState`
2. a `PoppedTlsState` for when it is popped, and this would have a `fn push(self) -> PushedTlsState`

It wouldn't be *too* much extra boilerplate, and would help ensure that these methods are called correctly.